osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs

```diff
@@ -1,19 +1,528 @@
+
+            osuBeatmap.HitObjects = hitObjects;
+
+            base.ApplyToBeatmap(beatmap);
+        }
+
+        private IEnumerable<double> generateBeats(IBeatmap beatmap, IReadOnlyCollection<OsuHitObject> origHitObjects)
+        {
+            var startTime = origHitObjects.First().StartTime;
+            var endObj = origHitObjects.Last();
+            var endTime = endObj switch
+            {
+                Slider slider => slider.EndTime,
+                Spinner spinner => spinner.EndTime,
+                _ => endObj.StartTime
+            };
+
+            var beats = beatmap.ControlPointInfo.TimingPoints
+                               .Where(x => Precision.AlmostBigger(endTime, x.Time))
+                               .SelectMany(tp =>
+                               {
+                                   var tpBeats = new List<double>();
+                                   var currentTime = tp.Time;
+
+                                   while (Precision.AlmostBigger(endTime, currentTime) && beatmap.ControlPointInfo.TimingPointAt(currentTime) == tp)
+                                   {
+                                       tpBeats.Add(Math.Floor(currentTime));
+                                       currentTime += tp.BeatLength;
+                                   }
+
+                                   return tpBeats;
+                               })
+                               .Where(x => Precision.AlmostBigger(x, startTime))
+                               // Remove beats during breaks
+                               .Where(x => !beatmap.Breaks.Any(b =>
+                                   Precision.AlmostBigger(x, b.StartTime)
+                                   && Precision.AlmostBigger(origHitObjects.First(y => Precision.AlmostBigger(y.StartTime, b.EndTime)).StartTime, x)
+                               ))
+                               .ToList();
+
+            // Remove beats that are too close to the next one (e.g. due to timing point changes)
+            for (var i = beats.Count - 2; i >= 0; i--)
+            {
+                var beat = beats[i];
+
+                if (Precision.AlmostBigger(beatmap.ControlPointInfo.TimingPointAt(beat).BeatLength / 2, beats[i + 1] - beat)) beats.RemoveAt(i);
+            }
+
+            return beats;
+        }
+
+        private void addHitSamples(IEnumerable<OsuHitObject> hitObjects, IReadOnlyList<OsuHitObject> origHitObjects)
+        {
+            var lastSampleIdx = 0;
+
+            foreach (var x in hitObjects)
+            {
+                var samples = getSamplesAtTime(origHitObjects, x.StartTime);
+
+                if (samples == null)
+                {
+                    while (lastSampleIdx < origHitObjects.Count && origHitObjects[lastSampleIdx].StartTime <= x.StartTime)
+                        lastSampleIdx++;
+                    lastSampleIdx--;
+
+                    if (lastSampleIdx < 0 && lastSampleIdx >= origHitObjects.Count) continue;
```

Comment:
What's the decrement for? And the `lastSampleIdx >= Count` check, too?

Reply:
The decrement was originally needed because the `while` always loops past the target hit object. I've reworked the logic a bit so that it is no longer needed.